KAFKA-15717: Added KRaft support in LeaderEpochIntegrationTest

appchemist opened the pull request:
This PR adds KRaft support in LeaderEpochIntegrationTest

JIRA : https://issues.apache.org/jira/projects/KAFKA/issues/KAFKA-15717

Test Result
![Screen_Shot_2024-01-18_At_8 57 23_Pm_1](https://github.com/apache/kafka/assets/1546031/b4dcfe88-11b0-40bc-91d1-784ac1428486)

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

appchemist committed:
KAFKA-15717: KRaft support in LeaderEpochIntegrationTest

ex172000 commented on core/src/test/scala/unit/kafka/server/epoch/LeaderEpochIntegrationTest.scala:
```diff
@@ -277,6 +283,22 @@ class LeaderEpochIntegrationTest extends QuorumTestHarness with Logging {
     producer.close()
   }
 
+  private def createTopicWithAssignment(topic: String, partitionReplicaAssignment: collection.Map[Int, Seq[Int]]): Unit = {
```
Is it possible to place this method inside the `TestUtils`?

appchemist replied:
@ex172000 Of course, I did.

appchemist replied:
@ex172000 
In `TestUtils`, How about changing the method names from `createTopicWithAssignment` to `createTopic`, and from `createTopic` to `createTopicWithZkClient`, similar to `createTopicWithAdmin`?

appchemist committed:
KAFKA-15717: KRaft support in LeaderEpochIntegrationTest

- move createTopicWithAssignment method to TestUtils.scala

appchemist commented:
@benstopford Please if you can review and provide the feedback.

mimaison commented on core/src/test/scala/unit/kafka/utils/TestUtils.scala:
```diff
@@ -559,6 +559,31 @@ object TestUtils extends Logging {
+   * If zkClient exists, create a topic in Zookeeper.
+   * Otherwise, create a topic using the Admin Client.
+   */
+  def createTopicWithAssignment(topic: String,
```
Could we reuse `createTopicWithAdmin()` instead of creating a new method? It's fine to use the Admin client to create the topic even in ZK mode.

appchemist replied:
Thank You!

mimaison commented on core/src/test/scala/unit/kafka/server/epoch/LeaderEpochIntegrationTest.scala:
```diff
@@ -142,17 +143,20 @@ class LeaderEpochIntegrationTest extends QuorumTestHarness with Logging {
-    brokers += createServer(fromProps(createBrokerConfig(100, zkConnect)))
-    assertEquals(100, TestUtils.waitUntilControllerElected(zkClient))
+    brokers += createBroker(fromProps(createBrokerConfig(100, zkConnectOrNull)))
+    if (!isKRaftTest()) {
```
Can we have a similar assertion in KRaft mode?

appchemist replied:
Of course, I will.
I checked the leader election using `KafkaRaftManager` in KRaft Mode. Is this a good approach?

mimaison reviewed (commented):
Thanks for the PR. I left a couple of comments.

appchemist committed:
KAFKA-15717: KRaft support in LeaderEpochIntegrationTest

- Use the Admin client to create the topic even in ZK mode.
- Add waitUntilLeaderElected in TestUtils

appchemist committed:
KAFKA-15717: KRaft support in LeaderEpochIntegrationTest

- Refactor waitUntilQuorumLeaderElected

mimaison reviewed (approved):
LGTM, thanks!

mimaison commented:
Thanks for the contribution!
Do you have a Jira account? If so I can assign https://issues.apache.org/jira/browse/KAFKA-15717 to you.

appchemist commented:
Thanks for the review! @mimaison
Yes, I have
About my jira account, Username and Fullname are appchemist.

mimaison commented:
Done, thanks
